### ISSUE TITLE
C# Invalid Hostname hangs

### ISSUE THREAD

**Issue by Azert-Mark** [grpc/grpc#9249]
Hi,

Sorry if this has been posted previously, I've had a search and can't find anything obviously similar.

I was writing a GRPC client using the C# libraries and noticed that an invalid Port number throws an RpcException of status Unavailable which is great. I then assumed that an invalid hostname would do the same, but it doesn't. Instead what it does is hangs indefinitely with no exception being thrown.

Is this expected behaviour? If so what is the correct way of checking invalid hostname, I have done the following, but do let me know if this is wrong:

```
            try
            {
                var channel = new Channel(instance.Host, instance.Port, ChannelCredentials.Insecure);
                channel.ConnectAsync(DateTime.Now.AddMilliseconds(instance.TimeoutMilliseconds).ToUniversalTime()).Wait();
                
                client = new SagaServer.SagaServerClient(channel);
            }
            catch (AggregateException ex)
            {
                if(ex.InnerException is TaskCanceledException)
                    throw new MyCustomException("My custom message");

                throw;
            }
```

**Comment by jtattermusch** [comment]
So the program doesn't actually hang, it just keeps retrying to resolve the invalid hostname repeatedly.

after setting the environment variable `GRPC_VERBOSITY=DEBUG` and enabling logging in gRPC C#
`GrpcEnvironment.SetLogger(new Grpc.Core.Logging.ConsoleLogger());`, you can see this output:

```
I0126 17:43:23.623398 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:191: dns resolution failed (will retry): {"created":"@1485449003.623299649","description":"OS Error","errno":-2,"file":"src/core/lib/iomgr/resolve_address_posix.c","file_line":115,"os_error":"Name or service not known","syscall":"getaddrinfo","target_address":"nonexistenthost.com:23456"}
D0126 17:43:23.623450 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:197: retrying in 1.000000000 seconds
I0126 17:43:24.624561 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:191: dns resolution failed (will retry): {"created":"@1485449004.624465594","description":"OS Error","errno":-2,"file":"src/core/lib/iomgr/resolve_address_posix.c","file_line":115,"os_error":"Name or service not known","syscall":"getaddrinfo","target_address":"nonexistenthost.com:23456"}
D0126 17:43:24.624619 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:197: retrying in 1.000000000 seconds
I0126 17:43:25.625591 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:191: dns resolution failed (will retry): {"created":"@1485449005.625488591","description":"OS Error","errno":-2,"file":"src/core/lib/iomgr/resolve_address_posix.c","file_line":115,"os_error":"Name or service not known","syscall":"getaddrinfo","target_address":"nonexistenthost.com:23456"}
D0126 17:43:25.625661 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:197: retrying in 1.000000000 seconds
I0126 17:43:26.626921 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:191: dns resolution failed (will retry): {"created":"@1485449006.626816840","description":"OS Error","errno":-2,"file":"src/core/lib/iomgr/resolve_address_posix.c","file_line":115,"os_error":"Name or service not known","syscall":"getaddrinfo","target_address":"nonexistenthost.com:23456"}
D0126 17:43:26.626988 140668953523968 src/core/ext/resolver/dns/native/dns_resolver.c:197: retrying in 1.000000000 seconds
```

Your code for detecting if it's possible to connect to the host seems reasonable (the code itself has some async antipatterns in it - instead, you should run in the async context and shouldn't use Wait(). But the idea is alright). Alternatively, you could assign a deadline to your calls (so that they would fail if once fails to resolve the name for too long). You can also start a probing call with a short deadline to see if you can connect to the backend.

Also, I believe that the retries for DNS resolver are configurable (through `ChannelOption`s?), but I am not sure how that's done. @dgquintas, are the DNS resolver retries configurable by the enduser?

**Comment by dgquintas** [comment]
No, the number of retries is not configurable. The knobs we currently have are for the backoff retry timer (initial delay, multiplier, etc) but those are not yet exposed through channel args.

I do find it surprising that the retry deadline is not increasing with every failure. Which version of core is this on?

**Comment by jtattermusch** [comment]
Should it ever time-out? It doesn't seem that is happening (unless the total timeout is very high).

**Comment by dgquintas** [comment]
No, it doesn't time out on its own. For as long as the call is alive, it'll continue to retry.

**Comment by Azert-Mark** [comment]
Hi @jtattermusch, thanks for getting back. I knew someone would point out not running it in an async context, this is due to this code running in a synchronous Castle Windsor installer, it doesn't have an async version unfortunately.

@dgquintas this is on the latest core currently available on Nuget, 1.0.1 at time of writing.

I guess my question is, should this try indefinitely and if it should, we need a way to limit either number of retries, or duration of retries. If the DNS doesn't exist at the start of the calls, the likelihood of it ever existing is quite slim I would assume.

**Comment by jtattermusch** [comment]
I created the issue https://github.com/grpc/grpc/issues/9481 to look more into DNS resolver behavior.
Closing this issue.
